### ISSUE TITLE
Multi-threading / RAM optimization in Constraint Optimization

### ISSUE THREAD

**Issue by olushako** [google/or-tools#788]
Dear community, 
I have faced with the next issue: I am running CVRP task in ortools with 3600 nodes. The app is using near 10% of CPU and only 16-20MB of RAM. 

My PC configuration: Core i7, 8 cores, 16 GB of RAM. 

if it possible to make sure what this issue is using more resources to find optimal results?

I have increased modelParameters.MaxCallbackCacheSize = 7200, but no change. 

My code:

`
var nr_vehicle = nb_trucks;

        var searchParameters = RoutingModel.DefaultSearchParameters();
        var modelParameters = RoutingModel.DefaultModelParameters();

        modelParameters.MaxCallbackCacheSize = 7200;

        var routing = new RoutingModel(unit.Count, nr_vehicle, 0, modelParameters);
        NodeEvaluator2 demandCallback = new Demand();
        routing.AddDimension(demandCallback, 0, maxCapacity, true, "capacity");

        
        NodeEvaluator2 lenghtcallback = new RouteLenght();
        routing.AddDimension(lenghtcallback, 0, maxCapacity, true, "routelenght");
        var dist_dimension = routing.GetDimensionOrDie("routelenght");
        for(long i=0; i<routing.Nodes();i++)
            dist_dimension.CumulVar(i).SetMax((long)Math.Truncate(Parameters.maxLength));

        NodeEvaluator2 nodesCount = new NodesCounter();
        routing.AddDimension(nodesCount, 0, maxCapacity, true, "nodescounter");
        var nodes_dimension = routing.GetDimensionOrDie("nodescounter");
        for (long i = 0; i < routing.Nodes(); i++)
            nodes_dimension.CumulVar(i).SetMax((long)Parameters.maxNodes-1);

        var distances = new RandomManhattan();
        routing.SetCost(distances);

        searchParameters.TimeLimitMs = Parameters.maxOptSearchTime;
        
        searchParameters.FirstSolutionStrategy = FirstSolutionStrategy.Types.Value.PathCheapestArc;

        var solution = routing.SolveWithParameters(searchParameters);`

**Comment by lperron** [comment]
The routing library is sequential.
I guess you could have moderate gains with // Local search, but I would not guarantee optimality.

What you can do is either decompose you problem, are partially assign visits to vehicles and run different this partial assignments in parallel.

We have no immediate plan to provide parallelism for the routing library, as for large problems, it is better to decompose upfront, and run each decomposition in parallel.

**Comment by Mizux** [comment]
duplicate of #417

**Comment by GuyBenhaim** [comment]
Hello,
Has there been any progress on this front?

Running VRP (with many nodes) using powerful Server different CPU cores participate, during different 'time-slices', but most of the time the cores are not used. Result is no real impriovement over single-core CPU.

Is there any way to structure the model such that inherent paralellizing actually becomes effective?

Of course, decomposing the problem and solving separately is possible but affects optimality.

Guy
